pkg/etw/event.go

```diff
@@ -0,0 +1,89 @@
+package etw
+
+// Channel represents the ETW logging channel that is used. It can be used by
+// event consumers to give an event special treatment.
+type Channel uint8
+
+const (
+	// ChannelTraceLogging is the default channel for TraceLogging events. It is
+	// not required to be used for TraceLogging, but will prevent decoding
+	// issues for these events on older operating systems.
+	ChannelTraceLogging Channel = 11
+)
+
+// Level represents the ETW logging level. There are several predefined levels
+// that are commonly used, but technically anything from 0-255 is allowed.
+// Lower levels indicate more important events, and 0 indicates an event that
+// will always be collected.
+type Level uint8
+
+// Predefined ETW log levels.
+const (
+	LevelAlways Level = iota
+	LevelCritical
+	LevelError
+	LevelWarning
+	LevelInfo
+	LevelVerbose
+)
+
+// Event represents a single ETW event. It can have field metadata and data
+// added to it, and then be logged via a provider to actually send it to ETW.
+type Event struct {
+	Descriptor *EventDescriptor
+	Metadata   *EventMetadata

```

Comment:
Is pointer to member necessary on all of these? Wondering if we can cut down on the number of allocations per event.

Also wondering if Descriptor needs to be public (and same for Metadata and Data pending discussion on my other comment).

Reply:
I made them pointers so you could e.g. replace Metadata with a pre-constructed metadata object, without needing a full copy. I don't have a good sense of what Go devs consider too large to copy around. If we don't care about the copy overhead, it should be just fine to get rid of the pointers here.

Reply:
Event is gone completely now. I'm still a bit curious as to what a good breakpoint would be on object size to change from pass-by-value to pass-by-pointer. Also, do you know if arrays are allocated on the heap or not? That could make a big difference for bytes.Buffer since it contains a [64]byte array.